src/pure/RealTimeQueue.mo

```diff
@@ -0,0 +1,1026 @@
+  ///   assert Queue.peekFront(queue) == ?1;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func peekFront<T>(queue : Queue<T>) : ?T = switch queue {
+    case (#empty) null;
+    case (#one(x)) ?x;
+    case (#two(x, _)) ?x;
+    case (#three(x, _, _)) ?x;
+    case (#idles((l, _), _)) Stacks.first(l);
+    case (#rebal((dir, big, small))) switch dir {
+      case (#left) ?SmallState.peek(small);
+      case (#right) ?BigState.peek(big)
+    }
+  };
+
+  /// Inspect the optional element on the back end of a queue.
+  /// Returns `null` if `queue` is empty. Otherwise, the back element of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushFront(Queue.pushFront(Queue.empty<Nat>(), 2), 1);
+  ///   assert Queue.peekBack(queue) == ?2;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func peekBack<T>(queue : Queue<T>) : ?T = switch queue {
+    case (#empty) null;
+    case (#one(x)) ?x;
+    case (#two(_, y)) ?y;
+    case (#three(_, _, z)) ?z;
+    case (#idles(_, (r, _))) Stacks.first(r);
+    case (#rebal((dir, big, small))) switch dir {
+      case (#left) ?BigState.peek(big);
+      case (#right) ?SmallState.peek(small)
+    }
+  };
+
+  /// Insert a new element on the front end of a queue.
+  /// Returns the new queue with `element` in the front followed by the elements of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushFront(Queue.pushFront(Queue.empty<Nat>(), 2), 1);
+  ///   assert Queue.peekFront(queue) == ?1;
+  ///   assert Queue.peekBack(queue) == ?2;
+  ///   assert Queue.size(queue) == 2;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)` worst-case!
+  ///
+  /// Space: `O(1)` worst-case!
+  public func pushFront<T>(queue : Queue<T>, element : T) : Queue<T> = switch queue {
+    case (#empty) #one(element);
+    case (#one(y)) #two(element, y);
+    case (#two(y, z)) #three(element, y, z);
+    case (#three(a, b, c)) {
+      let i1 = ((?(element, ?(a, null)), null), 2);
+      let i2 = ((?(c, ?(b, null)), null), 2);
+      #idles(i1, i2)
+    };
+    case (#idles(l0, (r, nR))) {
+      let (l, nL) = Idle.push(l0, element); // enque the element to the left end
+      // check if the size invariant still holds
+      if (3 * nR >= nL) {
+        debug assert 3 * nL >= nR;
+        #idles((l, nL), (r, nR))
+      } else {
+        // initiate the rebalancing process
+        let targetSizeL = nL - nR - 1 : Nat;
+        let targetSizeR = 2 * nR + 1;
+        debug assert targetSizeL + targetSizeR == nL + nR;
+        let big = #big1(Current.new(l, targetSizeL), l, null, targetSizeL);
+        let small = #small1(Current.new(r, targetSizeR), r, null);
+        let states = (#right, big, small);
+        let states6 = States.step(States.step(States.step(States.step(States.step(States.step(states))))));
```

Comment:
For future consideration: this might be a red-herring, but function calls are expensive on the IC. I wonder if using an iterative, tail-recursive version of `State.step(*n*, state) : state` could actually save some cost over apply State.step n times.
The body would be compiled to a loop and a jump which might just be cheaper, thought there will be some overhead due to checking the loop-counter.
Of course, inlining State.step would be even better but we don't do that.

Reply:
Actually it made it worse. 

|                            |   pure/Queue | pure/RealTimeQueue | mutable Queue |
| :------------------------- | -----------: | -----------------: | ------------: |
| Initialize with 2 elements |   3_571 (0%) |         2_742 (0%) |    3_401 (0%) |
| Push 500 elements          | 103_492 (0%) | 1_113_817 (+5.12%) |  243_585 (0%) |
| Pop front 2 elements       |  98_792 (0%) |         5_243 (0%) |    4_326 (0%) |
| Pop 150 front&back         | 106_545 (0%) |   453_232 (+2.93%) |  140_211 (0%) |

Both as a tail-rec function step and as an explicit while loop.

Reply:
Concurs with my experiments.